### PR TITLE
Optimization of the fallout_GPU subroutine

### PR DESCRIPTION
**Pull request by sherimickelson**

### CODE REVIEW

**Comment by sherimickelson** [comment]
[Optimize fallout_GPU on GPU](https://trello.com/c/eaxe6CUe/26-optimize-falloutgpu-on-gpu)

File: src/kessler.F [hunk 1]
```diff
@@ -736,80 +736,100 @@ subroutine fallout_GPU(dt,train,ruh,rvh,zh,mh,mf,rain,prate,rr,rho,   &
...
-      do k=(nk-1),1,-1
         if( q3d(i,j,k).lt.1.0e-8 ) vq(i,j,k) = vq(i,j,k+1)
       enddo
+    enddo
     enddo
       !!dtfall=dt/nfall(i) moved to up above
-    !$acc loop collapse(3) gang vector
+    !$acc parallel default(present) private(i,j,k,n)
+    !$acc loop gang vector collapse(4) 
+    do n=1,nfall(i,j)
```
Comment (johnmauff) [inline src/kessler.F:786]: I  don't think you are using the correct value for the bounds on the n-loop nfall(i,j) is using old values for i and j. here

File: src/kessler.F [hunk 2]
```diff
@@ -736,80 +736,100 @@ subroutine fallout_GPU(dt,train,ruh,rvh,zh,mh,mf,rain,prate,rr,rho,   &
...
         enddo
       enddo
     enddo
-    !$acc loop collapse(3) gang vector
+    enddo
+    !$acc end parallel
+
+    !$acc parallel default(present) private(i,j,k,n)
+    !$acc loop gang vector collapse(4) 
+    do n=1,nfall(i,j)
```
Comment (johnmauff) [inline src/kessler.F:800]: Another bug with loop bounds

File: src/kessler.F [hunk 3]
```diff
@@ -736,80 +736,100 @@ subroutine fallout_GPU(dt,train,ruh,rvh,zh,mh,mf,rain,prate,rr,rho,   &
...
         enddo
       enddo
     enddo
-    !$acc loop collapse(3) gang vector
+    enddo
+
+    !$acc end parallel
+    !$acc parallel default(present) private(i,j,n,nr)
+    !$acc loop gang vector collapse(4)
+    do n=1,nfall(i,j)
```
Comment (johnmauff) [inline src/kessler.F:814]: Another n-loop bounds error

File: src/kessler.F [hunk 4]
```diff
@@ -736,80 +736,100 @@ subroutine fallout_GPU(dt,train,ruh,rvh,zh,mh,mf,rain,prate,rr,rho,   &
...
                                 ! div by 1000 kg/m^3  (density of water)
         enddo
       enddo
     enddo
-    !$acc loop collapse(2) gang vector
+    enddo
+    !$acc end parallel
+    !$acc parallel default(present) private(i,j,n)
+    !$acc loop gang vector collapse(3) 
+    do n=1,nfall(i,j)
```
Comment (johnmauff) [inline src/kessler.F:828]: Another n-loop bounds error

**Review by johnmauff — CHANGES_REQUESTED** [review]
I think you have a bug here.  The bounds for the outer loop of n is nfall(i,j).  But i and j are not defined properly here.

**Comment by sherimickelson** [comment]
@johnmauff , Thanks for the catch.  I'll close this PR, fix it, and open a new PR.

**Pull request closed without merging**